### PR TITLE
feat: add openocean

### LINKED ISSUE #5107: [Bounty] Integrate OpenOcean, a multi-chain CEX/DEX aggregator into Mask Plugin
## Catalogue

- Platform
  - [x] Gitcoin
  - [ ] Hackerlink
  - [x] Mask
- Experience
  - [ ] Beginner
  - [x] Intermediate
  - [ ] Advanced
- Category
  - [ ] Mask Core Feature
  - [x] Mask Plugin
  - [ ] Security/Cryptography Feature
  - [ ] Documentation
  - [ ] Design
  - [ ] Other
    - [ ] /_List it here_/
- Time Commitment
  - [x] 2 weeks
- Bounty Size
  - [x] 150 $MASK

## Part 1: Problem Description

**Background**
OpenOcean is the world’s first full aggregation protocol for crypto trading that sources liquidity from DeFi and CeFi markets and enables cross-chain swaps (across Ethereum, Ethereum Layer2 such as Loopring and Polygon, Binance Smart Chain, Solana, Avalanche, Fantom, OEC, HECO, Ontology, TRON). Their intelligent routing algorithm finds the best prices from DEXes and CEXes and splits the routes to provide traders with the best prices with low slippage and fast settlement. OpenOcean provides an API and arbitrage tools for users to operate automated arbitrage strategies.

Mask Network has been seeking to provide more trading/swapping options for users, especially with trading protocols that provide multi-chain solutions. 

**Solution**
Please condense the OpenOcean pool & trading info into the existing design. 
- Integrate OpenOcean under the savings button on the entry page
- Create a plugin that integrates OpenOcean Swap features

See our mock design [here](https://www.figma.com/file/gVkQ67y285b4FXVV1KPThN/TwitterV1?node-id=17600%3A374185), does not need to be 1-to-1 but should follow the flow.

The UI for the swap plugin should be identical to other swaps
(On Polygon)
On Polygon

(On Ethereum mainnet)
On Ethereum mainnet

Any link cashtag with viable ticker, like $ETH on Twitter, could trigger the plugin. Swaps are also accessible via top line search bar and the swap entry on the left bar menu.

If you have ideas that might make more sense, please talk to us.

Required verbiage to help spread the word: “I just swapped xxx $[tokenname] into $[tokenname] with OpenOcean @OpenOceanGlobal on Mask extension @realMaskNetwork”

## Part 2: Resources

- [Bounty Guide](https://github.com/DimensionDev/Maskbook/tree/master/docs/bounty-guide.md)
- [OpenOcean](https://openocean.finance/)

## Part 3: Contact and Support
@LindaMask for ideas
@guanbinrui for implementations

Mask Network discord server: https://discord.gg/4Xccj7a75Q
OpenOcean discord server: https://discord.gg/cxK6CCYcGp 

Please speak up in the Mask discord's bounty section, so that we can find you on Discord. We can create a dedicated group for faster communication once we found each other

## Submission Requirements
Please include demo to show the integration is working.

## Judging Criteria
- Have a working demo
- Clean code & tech design
- Clear UI, easy to use, and generally good user experience

### PR DESCRIPTION
**Pull request by cformcn**
closes #5107

### CODE REVIEW

File: packages/mask/src/plugins/Trader/apis/openocean/index.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+            account: request.userAddr,
+        }),
+    )
+    const payload = await response.json()
+    // if (payload.status !== 200) {
+    //     throw new Error((payload as SwapOOErrorResponse).data ?? 'Unknown Error')
+    // }
+    const { data, outAmount, minOutAmount, to, value } = payload
+    const _resAmount = new BigNumber(outAmount).dividedBy(pow10(request.toToken.decimals ?? 0)).toNumber()
+    const _fromAmount = new BigNumber(request.fromAmount).dividedBy(pow10(request.fromToken.decimals ?? 0)).toNumber()
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Trader/apis/openocean/index.ts:27]: Using JS numbers for big integers will lead to overflow.

```suggestion
    const _resAmount = shiftLeft(outAmount, request.toToken.decimals).toFixed(),
    const _fromAmount = shiftLeft(request.fromAmount, request.fromToken.decimals).toFixed(),
```

File: packages/mask/src/plugins/Trader/apis/openocean/index.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        fromAmount: _fromAmount,
+        resPricePerFromToken: +(_fromAmount / _resAmount).toFixed(8),
+        resPricePerToToken: +(_resAmount / _fromAmount).toFixed(8),
+        to,
+        value,
+        slippage: request.slippage,
+        fromTokenSymbol: request.fromToken.symbol,
+        toTokenSymbol: request.toToken.symbol,
+        minOutAmount,
+    } as SwapOOData
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Trader/apis/openocean/index.ts:42]: `async function swapOO(request: SwapOORequest): SwapOOData { `

File: packages/mask/src/plugins/Trader/constants/openocean.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import { NetworkType } from '@masknet/web3-shared-evm'
+
+export const OPENOCEAN_BASE_URL = 'https://ethapi.openocean.finance/v2/'
+
+export const networkNames: Record<NetworkType, string> = {
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Trader/constants/openocean.ts:5]: We have a helper for this. Refer to `packages/web3-shared/evm/utils/enum.ts`.

File: packages/mask/src/plugins/Trader/apis/openocean/index.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+            account: request.userAddr,
+        }),
+    )
+    const payload = await response.json()
+    // if (payload.status !== 200) {
+    //     throw new Error((payload as SwapOOErrorResponse).data ?? 'Unknown Error')
+    // }
+    const { data, outAmount, minOutAmount, to, value } = payload
+    // const _resAmount = new BigNumber(outAmount).dividedBy(pow10(request.toToken.decimals ?? 0)).toNumber()
+    // const _fromAmount = new BigNumber(request.fromAmount).dividedBy(pow10(request.fromToken.decimals ?? 0)).toNumber()
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Trader/apis/openocean/index.ts:26]:
```suggestion
    const { data, outAmount, minOutAmount, to, value } = payload
```

**Review by septs — CHANGES_REQUESTED** [review]
plz use https://aka.pw/svgomg reduce svg images

File: packages/mask/src/resources/OpenOceanIcon.tsx [hunk 5]
```diff
@@ -0,0 +1,32 @@
+import { SvgIconProps, SvgIcon } from '@mui/material'
+
+const svg = (
+    <svg width="27" height="24" viewBox="0 0 27 24" version="1.1" xmlns="http://www.w3.org/2000/svg">
+        <title>Fill-1 Clipped</title>
+        <desc>Created with Sketch.</desc>
```
Comment (Jack-Works) [inline packages/mask/src/resources/OpenOceanIcon.tsx:6]: Please cleanup those properties.

**Comment by wenluomask** [comment]
1. Missing Unlock token button in BSC and Polygon. networks except for ETH. I tried to swap MASK->BNB, in this case the unlock button should show up;
Screen Shot 2021-12-27 at 11 58 40 AM
Screen Shot 2021-12-27 at 12 01 12 PM

2. For the network didn't connected, OpenOcean DEX option should be show up as well. For example, if network is Polygon, I cannot find OpenOcean in DEX list unless I change to BSC;
Screen Shot 2021-12-27 at 1 09 26 PM

3. Each DEX needs to have estimate gas fee and show in DEX car since the best option is chosen by MAX($(minReceived)-$(estimate gas fee)). OpenOcean doesn't have estimate gas fee for networks except for ETH;
Screen Shot 2021-12-27 at 1 06 36 PM

**Comment by hanszhoou** [comment]
@wenluomask Fixed.

**Comment by wenluomask** [comment]
#1 was not fixed yet.

**Comment by hanszhoou** [comment]
fixed

**Comment by wenluomask** [comment]
Checked.

**Pull request merged**